libscript/src/script-instance.cpp

```diff
@@ -2472,6 +2387,112 @@ bool MCScriptCallHandlerOfInstanceInternal(MCScriptInstanceRef self, MCScriptHan
     return t_success;
 }
 
+static uindex_t MCScriptComputeHandlerIndexOfInstance(MCScriptInstanceRef p_instance, MCScriptCommonHandlerDefinition *p_handler)
+{
+    uindex_t t_min, t_max;
+    t_min = 0;
+    t_max = p_instance -> handler_count;
+    while(t_min < t_max)
+    {
+        uindex_t t_mid;
+        t_mid = (t_min + t_max) / 2;
+        
+        if (p_instance -> handlers[t_mid] . definition < p_handler)
+            t_min = t_mid + 1;
+        else
+            t_max = t_mid;
+    }
+    
+    return t_min;
+}
+
+bool MCScriptEvaluateHandlerOfInstanceInternal(MCScriptInstanceRef p_instance, MCScriptCommonHandlerDefinition *p_handler, MCHandlerRef& r_handler)
+{
+    // Compute the index in the handler table of p_handler; then, if it is the
+    // definition we are looking for, return its previously computed value.
+    uindex_t t_index;
+    t_index = MCScriptComputeHandlerIndexOfInstance(p_instance, p_handler);
+    if (t_index < p_instance -> handler_count &&
+        p_instance -> handlers[t_index] . definition == p_handler)
+    {
+        r_handler = p_instance -> handlers[t_index] . value;
+        return true;
+    }
+    
+    // Calculate the handlerref value we need.
+    MCHandlerRef t_value;
+    if (p_handler -> kind == kMCScriptDefinitionKindHandler)
+    {
+        // LCB handlers are easy - we just wrap up the instance and handler definition
+        // pair in a handler-ref.
+        
+        MCScriptHandlerDefinition *t_handler_definition;
+        t_handler_definition = static_cast<MCScriptHandlerDefinition *>(p_handler);
+        
+        MCTypeInfoRef t_signature;
+        t_signature = p_instance -> module -> types[t_handler_definition -> type] -> typeinfo;
+        
+        // The context struct is 'moved' into the handlerref.
+        __MCScriptHandlerContext t_context;
+        t_context . instance = MCScriptRetainInstance(p_instance);
```

Comment:
This retain is preventing script instances with handler objects being destroyed. I think MCScriptReleaseInstance should go through its handler array releasing any instances which have been retiained in the context.